### PR TITLE
Nwm v2 reanalysis

### PR DESCRIPTION
**Pull request by mohamedmorsy**
This pull request includes the following:
- Fixed minor bug within the method to retrieve the NWM reanalysis through AWS
- Added a capability to interact with and retrieve the NWM V2 reanalysis through GCS
- Added a use case to compare the NWM outputs with the USGS available observation
- Added new functions in the provided use case to interact with and retrieve the USGS observation (discharge and stage). These new functions might be moved to the proper location in the core files.

### CODE REVIEW

File: reanalysis/reanalysis.py [hunk 1]
```diff
@@ -99,3 +103,38 @@ def plotReanalysis(df: pd.DataFrame, comid: int, freq: str, flow: bool = True):
...
+    elif target_analysis == 'long_range':
+        target_dir = subfolders[1]
+    else:
+        raise('unavailable target analysis directory. Please double check!')
+        
+    records_wanted = list(pd.date_range(start_date, end_date, freq=freq).strftime('%Y%m%d%H%M'))
+    gs_paths = []
+    for record in records_wanted:
+        gs_paths.append(target_dir + record[:4] + "/" + record +'.CHRTOUT_DOMAIN1.comp')
+    return(gs_paths)
```
Comment (albrazeau) [inline reanalysis/reanalysis.py]: `return` is not a function, it is a statement. therefore, you should not have parentheses here

File: reanalysis/reanalysis.py [hunk 2]
```diff
@@ -99,3 +103,38 @@ def plotReanalysis(df: pd.DataFrame, comid: int, freq: str, flow: bool = True):
...
+        all_data = xr.open_mfdataset(openfiles)
+        print(round((time()-start), 2), 'seconds to access the data for the given duration')
+    if gs_access_method == 'download':
+        for file_path in file_paths_list:
+            fs.download(file_path,download_dir+ "/" + file_path.split("/")[-1])
+        files = glob(download_dir + '/*')
+        assert len(files) == len(file_paths_list), 'Downloading error'
+        all_data = xr.open_mfdataset(files)
+        print(round((time()-start), 2), 'seconds to download', sum([os.path.getsize(f) for f in files])/1e9, 'GB of data for the given duration')
+    return(all_data)
```
Comment (albrazeau) [inline reanalysis/reanalysis.py]: Same as above.

File: reanalysis/reanalysis.py [hunk 3]
```diff
@@ -99,3 +103,38 @@ def plotReanalysis(df: pd.DataFrame, comid: int, freq: str, flow: bool = True):
...
+
+def data_access_gs(file_paths_list, gs_access_method='real-time', download_dir='./data'):
+    start = time()
+    fs = gcsfs.GCSFileSystem(project='national-water-model-v2')
+    if gs_access_method == 'real-time':
+        openfiles = [fs.open(f, "rb") for f in file_paths_list]
+        all_data = xr.open_mfdataset(openfiles)
+        print(round((time()-start), 2), 'seconds to access the data for the given duration')
+    if gs_access_method == 'download':
+        for file_path in file_paths_list:
```
Comment (albrazeau) [inline reanalysis/reanalysis.py]: There is potential to make this download happen in parallel using pool very easily, if you are so inclined.

File: reanalysis/reanalysis.py (hunk 1 above)
Comment (mohamedmorsy) [reply reanalysis/reanalysis.py]: Done!

File: reanalysis/reanalysis.py (hunk 2 above)
Comment (mohamedmorsy) [reply reanalysis/reanalysis.py]: Done!

File: reanalysis/reanalysis.py (hunk 3 above)
Comment (mohamedmorsy) [reply reanalysis/reanalysis.py]: Done!

**Comment by mohamedmorsy** [comment]
@albrazeau This pull request is ready for you to review. Thanks!

File: reanalysis/reanalysis.py [hunk 4]
```diff
@@ -131,10 +137,11 @@ def data_access_gs(file_paths_list, gs_access_method='real-time', download_dir='
         all_data = xr.open_mfdataset(openfiles)
         print(round((time()-start), 2), 'seconds to access the data for the given duration')
     if gs_access_method == 'download':
-        for file_path in file_paths_list:
-            fs.download(file_path,download_dir+ "/" + file_path.split("/")[-1])
+        p = Pool(int(cpu_count()*2))
```
Comment (albrazeau) [inline reanalysis/reanalysis.py]: Use `with` here:
```python
with Pool(int(cpu_count()*2)) as p:
    p.map(single_dl, file_paths_list)
```

That way you don't need to do `p.close()` at the end... but more importantly, if the download fails for any reason (i.e. internet goes out), you won't leave processes stranded and unable to be closed. Using `with` will close them no matter what.

File: reanalysis/reanalysis.py [hunk 5]
```diff
@@ -120,7 +120,13 @@ def get_reanalysis_paths_gs(start_date, end_date, freq, target_analysis):
     gs_paths = []
     for record in records_wanted:
         gs_paths.append(target_dir + record[:4] + "/" + record +'.CHRTOUT_DOMAIN1.comp')
-    return(gs_paths)    
+    return gs_paths    
+
+
+def single_dl(file_path, download_dir = './data'):
```
Comment (albrazeau) [inline reanalysis/reanalysis.py]: Can you rename this function and the one above (line 74)? Specify that one is for s3 and one is for gcs.

File: reanalysis/reanalysis.py (hunk 4 above)
Comment (albrazeau) [reply reanalysis/reanalysis.py]: Can you also implement this in the `s3download_parallel` function in line 84 as well? Thanks!

File: reanalysis/reanalysis.py (hunk 5 above)
Comment (mohamedmorsy) [reply reanalysis/reanalysis.py]: Done!

File: reanalysis/reanalysis.py (hunk 4 above)
Comment (mohamedmorsy) [reply reanalysis/reanalysis.py]: Good point! Done!

**Pull request merged**